subsys/bluetooth/audio/has.c

```diff
@@ -77,45 +483,97 @@ static ssize_t write_control_point(struct bt_conn *conn, const struct bt_gatt_at
+		return -EINVAL;
+	}
+
+	CHECKIF(param->name == NULL) {
+		BT_ERR("param->name is NULL");
+		return -EINVAL;
+	}
+
+	name_len = strlen(param->name);
+
+	CHECKIF(name_len < BT_HAS_PRESET_NAME_MIN) {
+		BT_ERR("param->name is too short (%zu < %u)", name_len, BT_HAS_PRESET_NAME_MIN);
+		return -EINVAL;
+	}
+
+	CHECKIF(name_len > BT_HAS_PRESET_NAME_MAX) {
+		BT_WARN("param->name is too long (%zu > %u)", name_len, BT_HAS_PRESET_NAME_MAX);
+	}
+
+	preset_foreach(param->index, param->index, preset_found, &preset);
+	if (preset != NULL) {
+		return -EALREADY;
+	}
+
+	preset = preset_alloc(param->index, param->properties, param->name);
+	if (preset == NULL) {
+		return -ENOMEM;
+	}
+
+	return 0;
 }
 
-static void ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+int bt_has_preset_unregister(uint8_t index)
 {
-	BT_DBG("attr %p value 0x%04x", attr, value);
+	struct bt_has_preset *preset = NULL;
+
+	CHECKIF(index == BT_HAS_PRESET_INDEX_NONE) {
+		BT_ERR("index is invalid");
+		return -EINVAL;
+	}
+
+	preset_foreach(index, index, preset_found, &preset);
+	if (preset == NULL) {
+		return -ENOENT;
+	}
+
+	preset_free(preset);
+
+	return 0;
 }
-#endif /* CONFIG_BT_HAS_PRESET_COUNT > 0 */
 
-/* Hearing Access Service GATT Attributes */
-BT_GATT_SERVICE_DEFINE(has_svc,
-	BT_GATT_PRIMARY_SERVICE(BT_UUID_HAS),
-	BT_GATT_CHARACTERISTIC(BT_UUID_HAS_HEARING_AID_FEATURES, BT_GATT_CHRC_READ,
-			       BT_GATT_PERM_READ_ENCRYPT, read_features, NULL, NULL),
-#if CONFIG_BT_HAS_PRESET_COUNT > 0
-	BT_GATT_CHARACTERISTIC(BT_UUID_HAS_PRESET_CONTROL_POINT,
-#if defined(CONFIG_BT_EATT)
-			       BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE | BT_GATT_CHRC_NOTIFY,
-#else
-			       BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE,
-#endif /* CONFIG_BT_EATT */
-			       BT_GATT_PERM_WRITE_ENCRYPT, NULL, write_control_point, NULL),
-	BT_GATT_CCC(ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_ENCRYPT),
-	BT_GATT_CHARACTERISTIC(BT_UUID_HAS_ACTIVE_PRESET_INDEX,
-			       BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY, BT_GATT_PERM_READ_ENCRYPT,
-			       read_active_preset_index, NULL, NULL),
-	BT_GATT_CCC(ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_ENCRYPT),
+struct bt_has_preset_foreach_data {
+	bt_has_preset_func_t func;
+	void *user_data;
+};
+
+static uint8_t bt_has_preset_foreach_func(const struct bt_has_preset *preset, void *user_data)
+{
+	const struct bt_has_preset_foreach_data *data = user_data;
+
+	return data->func(preset->index, preset->properties, preset->name, data->user_data);
+}
+
+void bt_has_preset_foreach(uint8_t index, bt_has_preset_func_t func, void *user_data)
+{
+	uint8_t start_index, end_index = index;
+	struct bt_has_preset_foreach_data data = {
+		.func = func,
+		.user_data = user_data,
+	};
+
+	if (index == BT_HAS_PRESET_INDEX_NONE) {
+		start_index = BT_HAS_PRESET_INDEX_FIRST;
+		end_index = BT_HAS_PRESET_INDEX_LAST;
+	}
+
+	preset_foreach(start_index, end_index, bt_has_preset_foreach_func, &data);
```

Comment:
if `index != BT_HAS_PRESET_INDEX_NONE` then `start_index` is uninitialized. Suggest to do
```suggestion
	uint8_t start_index, end_index;
	struct bt_has_preset_foreach_data data = {
		.func = func,
		.user_data = user_data,
	};

	if (index == BT_HAS_PRESET_INDEX_NONE) {
		start_index = BT_HAS_PRESET_INDEX_FIRST;
		end_index = BT_HAS_PRESET_INDEX_LAST;
	} else {
		start_index = end_index = index
	}

	preset_foreach(start_index, end_index, bt_has_preset_foreach_func, &data);
```

Reply:
oh, thanks for the catch!